src/Tests/Tests/Test_FastConv.cpp

```diff
@@ -18,22 +19,262 @@ namespace fastconv_test {
+        memset(input, 0, sizeof(float) * signalLength); // an impulse as input signal at sample index 3
+        input[shift] = 1.F;
+
+        // initialize output signal
+        float* output = new float[signalLength];
+        memset(output, 0, sizeof(float) * signalLength);
+
+        // convolve with varying block size
+        int blockLengths[] = {
+            1, 13, 1023, 2048, 1, 17, 5000, 1897
+        };
+        for (int i = 0, accu = 0; i < 8; accu += blockLengths[i++]) {
+            m_pCFastConv->process(output + accu, input + accu, blockLengths[i]);
+        }
+
+        // check if the output signal is exactly the same with the impulse response
+        for (int i = 0; i < IRLength && i + shift < signalLength; ++i) {
+            EXPECT_NEAR(IR[i], output[i + shift], 1e-8);
+        }
+
+        // also check if the first three are ok
+        for (int i = 0; i < shift; ++i) {
+            EXPECT_EQ(output[i], 0);
+        }
+
+        // also check if tails are all zero
+        for (int i = IRLength + shift; i < signalLength; ++i) {
+            EXPECT_EQ(output[i], 0);
+        }
+
+        delete[] input;
+        delete[] output;
+    }
+
+    class FFTFastConv: public testing::Test
     {
+    protected:
+        void SetUp() override
+        {
+            m_pCFastConv = new CFastConv();
+
+            // generate a random IR of length 51 samples
+            IRLength = 51;
+            IR = new float[IRLength];
+            memset(IR, 0, sizeof(float) * IRLength);
+
+            // use mt19937 & distribution for high quality random number generation
+            std::mt19937 generator(114514); // 114514 is seed
+            std::uniform_real_distribution<float> uniformRealDistribution(-1.F, 1.F);
+            for (int i = 0; i < IRLength; ++i) {
+                IR[i] = uniformRealDistribution(generator);
+            }
+
+            m_pCFastConv->init(IR, IRLength, 32, CFastConv::kFreqDomain);
+        }
+
+        virtual void TearDown()
+        {
+            delete m_pCFastConv;
+            delete[] IR;
+        }
+
+        int IRLength = 0;
+        float* IR = nullptr;
+        CFastConv *m_pCFastConv = nullptr;
+    };
+
+    TEST_F(FFTFastConv, identity)
+    {
+        // initialize input signal, which is $\delta[n-3]$
+        int shift = 3;
+        int signalLength = 10;
+        float* input = new float[signalLength]; // make the input signal 10 samples long
+        memset(input, 0, sizeof(float) * signalLength); // an impulse as input signal at sample index 3
+        input[shift] = 1.F;
+
+        // initialize output signal
+        float* output = new float[signalLength];
+        memset(output, 0, sizeof(float) * signalLength);
+
+        m_pCFastConv->process(output, input, signalLength);
+
+        // check if the output signal is exactly the same with the impulse response
+        for (int i = 0; i + shift < signalLength; ++i) {
+            EXPECT_NEAR(IR[i], output[i + shift], 1e-6);
```

Comment:
hm, shouldn't the output have an additional latency of the block size (in this case, 32 samples)?

Reply:
We pre-allocate memory for future convolution results, rather than wait until there's enough sample. Therefore as long as there's some input, we can immediately return output as all past information that will be used to generate output in this call has already been written in the current output buffer region in previous calls. Therefore we don't have latency for this.